lib/media/quality_observer.js

```diff
@@ -0,0 +1,276 @@
+      if (mediaQualityAtPosition &&
+          !shaka.media.QualityObserver.mediaQualitiesAreTheSame_(
+              contentMediaQuality.currentQuality, mediaQualityAtPosition)) {
+        if (this.positionIsBuffered_(
+            positionInSeconds, mediaQualityAtPosition.contentType)) {
+          contentMediaQuality.currentQuality = mediaQualityAtPosition;
+          this.onQualityChange_(mediaQualityAtPosition);
+        }
+      }
+    }
+  }
+
+  /**
+   * Determine if a position is buffered for a given content type.
+   *
+   * @param {!number} position
+   * @param {!string} contentType
+   * @private
+   */
+  positionIsBuffered_(position, contentType) {
+    const bufferedInfo = this.getBufferedInfo_();
+    const bufferedRanges = bufferedInfo[contentType];
+    if (bufferedRanges && bufferedRanges.length > 0) {
+      const bufferStart = bufferedRanges[0].start;
+      const bufferEnd = bufferedRanges[bufferedRanges.length - 1].end;
+      if (position >= bufferStart && position < bufferEnd) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Removes the QualityChange(s) that are not relevant to the buffered
+   * content of the specified contentType. Note that this function is
+   * invoked just before adding the quality change event associated with
+   * the next media segment to be appended.
+   *
+   * @param {!string} contentType
+   * @private
+   */
+  purgeQualityChangesForContentType_(contentType) {
+    const bufferedInfo = this.getBufferedInfo_();
+    const contentMediaQuality = this.getContentMediaQuality_(contentType);
+    const bufferedRanges = bufferedInfo[contentType];
+
+    if (bufferedRanges && bufferedRanges.length > 0) {
+      const bufferStart = bufferedRanges[0].start;
+      const bufferEnd = bufferedRanges[bufferedRanges.length - 1].end;
+      const oldQualityChanges = contentMediaQuality.qualityChanges;
+      contentMediaQuality.qualityChanges =
+        oldQualityChanges.filter(
+            (qualityChange, index) => {
+              // Remove all but last quality change before bufferStart.
+              if ((qualityChange.position <= bufferStart) &&
+                (index + 1 < oldQualityChanges.length) &&
+                (oldQualityChanges[index + 1].position <= bufferStart)) {
+                return false;
+              }
+              // Remove all quality changes after bufferEnd.
+              if (qualityChange.position >= bufferEnd) {
+                return false;
+              }
+              return true;
+            });
+    } else {
+      // Nothing is buffered; so remove all quality changes.
+      contentMediaQuality.qualityChanges = [];
+    }
+  }
+};
+
+/**
+ * When media quality changes for audio or video
+ * content types, the onQualityChange listener is invoked
+ * with the new quality info.
+ *
+ * @typedef {function(shaka.extern.MediaQualityInfo)}
+ */
+shaka.media.QualityObserver.EventListener;
```

Comment:
It might be a good idea to also pass the position into this function.

Reply:
done